src/SpectatorView.Unity/Assets/SpectatorView/Scripts/StateSynchronization/AssetCache.cs

```diff
@@ -178,83 +220,158 @@ public static IEnumerable<T> EnumerateAllComponentsInScenesAndPrefabs<T>()
+        const string assetContentExtension = ".asset";
 
-        private Dictionary<AssetId, TAssetEntry> lookupByAssetId;
-        private Dictionary<TAsset, TAssetEntry> lookupByAsset;
+#pragma warning disable 414
+        [SerializeField]
+        private NameEntry[] assets = null;
+#pragma warning restore 414
 
-        protected IDictionary<AssetId, TAssetEntry> LookupByAssetId
+        private Dictionary<string, List<AssetId>> LookupByName
         {
             get
             {
-                if (lookupByAssetId == null)
+                if (lookupByName == null)
                 {
-                    if (assets == null)
-                    {
-                        lookupByAssetId = new Dictionary<AssetId, TAssetEntry>();
-                    }
-                    else
+                    lookupByName = new Dictionary<string, List<AssetId>>();
+                    if (assets != null)
                     {
-                        lookupByAssetId = assets.Where(a => a.Asset != null).ToDictionary(a => a.AssetId);
+                        for (int i = 0; i < assets.Length; i++)
+                        {
+                            lookupByName[assets[i].Name] = assets[i].Ids.ToList();
+                        }
                     }
                 }
-                return lookupByAssetId;
-            }
-        }
 
-        protected IDictionary<TAsset, TAssetEntry> LookupByAsset
-        {
-            get
-            {
-                if (lookupByAsset == null)
-                {
-                    if (assets == null)
-                    {
-                        lookupByAsset = new Dictionary<TAsset, TAssetEntry>();
-                    }
-                    else
-                    {
-                        lookupByAsset = assets.Where(a => a.Asset != null).ToDictionary(a => a.Asset);
-                    }
-                }
-                return lookupByAsset;
+                return lookupByName;
             }
         }
 
+#pragma warning disable 414
+        private Dictionary<string, List<AssetId>> lookupByName;
+        private Dictionary<AssetId, AssetCacheEntry> lookupByAssetId = new Dictionary<AssetId, AssetCacheEntry>();
+        private Dictionary<TAsset, AssetCacheEntry> lookupByAsset = new Dictionary<TAsset, AssetCacheEntry>();
+#pragma warning restore 414
+
         public TAsset GetAsset(AssetId assetId)
         {
             using (StateSynchronizationPerformanceMonitor.Instance.IncrementEventDuration(this.GetType().Name, "GetAsset"))
             {
-                TAssetEntry assetEntry;
-                if (LookupByAssetId.TryGetValue(assetId, out assetEntry))
+                if (assetId == null ||
+                assetId == AssetId.Empty ||
+                assetId.Name == string.Empty)
+                {
+                    return default(TAsset);
+                }
+
+                if (lookupByAssetId.TryGetValue(assetId, out var entry))

```

Comment:
For the performance monitor (again, I don't have a lot of background here, so maybe this question isn't fully valid), is it interesting to bucketize based on cache misses and cache hits? i.e. is the event duration in total interesting, or are we really interested in the real load time?

Reply:
we were just using the event duration in total to understand whether or not this new deferred asset loading strategy had a large impact on the app responsiveness compared to the old asset loading method. I think as more time passes we'll have to make this performance reporting better, but for now this event has served its purpose.

---
In reply to: [355040093](https://github.com/microsoft/MixedReality-SpectatorView/pull/295#discussion_r355040093) [](ancestors = 355040093)